gcloud-java-compute/src/main/java/com/google/gcloud/compute/AddressInfo.java

```diff
@@ -0,0 +1,576 @@
+     */
+    public abstract List<ResourceId> users();
+
+    final boolean baseEquals(Usage usage) {
+      return Objects.equals(toPb(), usage.toPb());
+    }
+
+    Address toPb() {
+      return new Address().setUsers(Lists.transform(users(), new Function<ResourceId, String>() {
+        @Override
+        public String apply(ResourceId resourceId) {
+          return resourceId.selfLink();
+        }
+      }));
+    }
+
+    @SuppressWarnings("unchecked")
+    static <T extends Usage> T fromPb(Address addressPb) {
+      String url = addressPb.getUsers().get(0);
+      if (InstanceId.matchesUrl(url)) {
+        return (T) InstanceUsage.fromPb(addressPb);
+      } else if (RegionForwardingRuleId.matchesUrl(url)) {
+        return (T) RegionForwardingUsage.fromPb(addressPb);
+      } else {
+        return (T) GlobalForwardingUsage.fromPb(addressPb);
+      }
+    }
+  }
+
+  /**
+   * Usage information for a Google Compute Engine region address assigned to a virtual machine
+   * instance.
+   */
+  public static final class InstanceUsage extends Usage {
+
+    private static final long serialVersionUID = -5028609518171408695L;
+
+    private final InstanceId instance;
+
+    InstanceUsage(InstanceId instance) {
+      this.instance = checkNotNull(instance);
+    }
+
+    /**
+     * Returns the identity of the instance using the address.
+     */
+    public InstanceId instance() {
+      return instance;
+    }
+
+    @Override
+    public List<ResourceId> users() {
+      return ImmutableList.<ResourceId>of(instance);
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this).add("instance", instance).toString();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      return obj instanceof InstanceUsage && baseEquals((InstanceUsage) obj);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(instance);
+    }
+
+    @SuppressWarnings("unchecked")
+    static InstanceUsage fromPb(Address addressPb) {
+      return new InstanceUsage(InstanceId.fromUrl(addressPb.getUsers().get(0)));
+    }
+  }
+
+  /**
+   * Usage information for a Google Compute Engine region address assigned to one or more region
+   * forwarding rules.
+   */
+  public static final class RegionForwardingUsage extends Usage {
+
+    private static final long serialVersionUID = -4255145869626427363L;
+
+    private final List<RegionForwardingRuleId> forwardingRules;
+
+    RegionForwardingUsage(List<RegionForwardingRuleId> forwardingRules) {
+      this.forwardingRules = ImmutableList.copyOf(forwardingRules);
+    }
+
+    /**
+     * Returns a list of identities of region forwarding rules that are currently using the address.
+     */
+    public List<RegionForwardingRuleId> forwardingRules() {
+      return forwardingRules;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public List<ResourceId> users() {
+      return (List<ResourceId>) (List) forwardingRules;
```

Comment:
Nasty, but I don't think the extra `(List<ResourceId>)` casting is necessary.
Also, if you take the suggestion above may any of this casting will be necessary.